File: solution/1530. Number of Good Leaf Nodes Pairs/1530.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countPairs(TreeNode* root, int distance) {
        int res = 0;
        dfs(root, distance, res);
        return res;
    }
    
private:
    vector<int> dfs(TreeNode* node, int distance, int& res) {
        vector<int> count(distance + 1);
        if (!node) return count;
        
        auto l = dfs(node->left, distance, res);
        auto r = dfs(node->right, distance, res);
        
        for (int i = 0; i < distance; ++i)
            for (int j = 0; i + j + 2 <= distance; ++j)
                res += l[i] * r[j];
        
        for (int i = 0; i < distance; ++i)
            count[i + 1] = l[i] + r[i];
        
        if (!node->left && !node->right)
            count[0] = 1;
        
        return count;
    }
};

File: solution/1530. Number of Good Leaf Nodes Pairs/1530.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countPairs(self, root: TreeNode, distance: int) -> int:
        res = 0
        
        def dfs(node):
            nonlocal res
            count = [0] * (distance + 1)
            if not node:
                return count
            
            l = dfs(node.left)
            r = dfs(node.right)
            
            for i in range(distance):
                for j in range(distance):
                    if i + j + 2 > distance:
                        break
                    res += l[i] * r[j]
            
            for i in range(distance):
                count[i + 1] = l[i] + r[i]
            
            if not node.left and not node.right:
                count[0] = 1
            
            return count
        
        dfs(root)
        return res
